src/_rollout.sh

```diff
@@ -0,0 +1,129 @@
+# Credits
+# "SchemaVersion": "0.1.0",
+# "Vendor": "Karol Musur",
+# "Version": "v0.5",
+# "ShortDescription": "Rollout new Compose service version"
+# "Adapted by": "Michael Burke"
+
+# Defaults
+HEALTHCHECK_TIMEOUT=60
+NO_HEALTHCHECK_TIMEOUT=10
+
+healthcheck() {
+  local container_id="$1"
+
+  if docker inspect --format='{{json .State.Health.Status}}' "$container_id" | grep -v "unhealthy" | grep -q "healthy"; then
+    return 0
+  fi
+
+  return 1
+}
+
+scale() {
+  local service="$1"
+  local replicas="$2"
+  compose_client up --detach --scale $service=$replicas --no-recreate "$service"
+}
+
+mod_rollout() {
+  # Added removal of the couchbase-migrations container due to this not getting attached to the new network scaled
+  if [ $(docker compose ps -a --format json | jq -re '.[].Name' | grep couchbase-migrations) != "" ]
+    then
+      debug "Removing 'couchbase-migrations' container"
+      docker rm -f `docker compose ps -a --format json | jq -re '.[].Name' | grep couchbase-migrations` > /dev/null 2>&1
+  fi
+  # Get list of services from Docker Compose Config
+  service_list=(
+    "datalake-maintainer"
+    "notification-engine"
+    "notification-sender"
+    "plextracapi"
+  )
+  #service_list=$(for i in `docker compose config --format json | jq -re '.services|keys[]'`; do \
+  #  docker compose config --format json | jq -re --arg v "$i" '.services | "\($v)=\(.[$v].image)"'; \
+  #  done | grep plextracapi | sort -u | cut -d '=' -f1)
+  debug "$service_list"
+  for s in ${service_list[@]}
+  #for s in `echo -ne "$service_list"`
+    do
+      SERVICE=$s
+      SCALE=$(docker compose config --format json | jq -re --arg v "$SERVICE" '.services | .[$v].deploy.replicas | select(. != null)')
+      if [ $SCALE == 0 ]
+        then
+          debug "$SERVICE show $SCALE replicas; skipping"
+          continue
+      fi
+      if [[ "$(compose_client ps --quiet "$SERVICE")" == "" ]]
+        then
+          debug "Service '$SERVICE' is not running. Starting the service."
+          compose_client up --detach --no-recreate "$SERVICE"
+          debug "$SERVICE created"
+          continue
+      fi
+
+      OLD_CONTAINER_IDS_STRING=$(compose_client ps --quiet "$SERVICE")
+      readarray -t OLD_CONTAINER_IDS <<<"$OLD_CONTAINER_IDS_STRING"
+
+      if [ $SCALE != "0" ]
+        then
+          SCALE_TIMES_TWO=$((SCALE * 2))
+        else
+          SCALE_TIMES_TWO=0
+          break
+      fi
+      debug "Scaling '$SERVICE' to '$SCALE_TIMES_TWO'"
+      scale "$SERVICE" $SCALE_TIMES_TWO
+
+      # create a variable that contains the IDs of the new containers, but not the old ones
+      NEW_CONTAINER_IDS_STRING=$(compose_client ps --quiet "$SERVICE" | grep --invert-match --file <(echo "$OLD_CONTAINER_IDS_STRING"))
+      readarray -t NEW_CONTAINER_IDS <<<"$NEW_CONTAINER_IDS_STRING"
```

Comment:
If I kill the rollout in the middle of the process (not normal behavior, but could possibly happen with end users) with a ctrl+c, it then fails to update or rollout and seems to require a `docker-compose down && docker-compose up` to become usable again. I think the issue is when the array is empty because there are fewer than expected container ids. Might be a use case worth investigating. Happy to show you what I'm seeing and help come up with a solution.

```
plextrac@vagrant:~$ plextrac rollout -v

______ _         _____              
| ___ \ |       |_   _|             
| |_/ / | _____  _| |_ __ __ _  ___ 
|  __/| |/ _ \ \/ / | '__/ _\ |/ __|
| |   | |  __/>  <| | | | (_| | (__ 
\_|   |_|\___/_/\_\_/_|  \__,_|\___|
                                    

Instance Management Utility v0.4.2

    Logged event 'command:mod_rollout' at 1696955106
    Event Details:
      activity: command:mod_rollout
      timestamp: Tue 10 Oct 2023 04:25:06 PM UTC
      user: plextrac
      data: -
    datalake-maintainer
    datalake-maintainer show 0 replicas; skipping
    docker-compose flags: ps --quiet notification-engine
    docker-compose configs:  -f /opt/plextrac/docker-compose.yml -f /opt/plextrac/docker-compose.override.yml
    docker-compose flags: ps --quiet notification-engine
    docker-compose configs:  -f /opt/plextrac/docker-compose.yml -f /opt/plextrac/docker-compose.override.yml
    Scaling 'notification-engine' to '2'
    docker-compose flags: up --detach --scale notification-engine=2 --no-recreate notification-engine
    docker-compose configs:  -f /opt/plextrac/docker-compose.yml -f /opt/plextrac/docker-compose.override.yml
[+] Building 0.0s (0/0)                                                                                                                                                                                                                        docker:default
[+] Running 4/0
 ✔ Container redis                            Running                                                                                                                                                                                                    0.0s 
 ✔ Container plextrac-plextracdb-1            Running                                                                                                                                                                                                    0.0s 
 ✔ Container plextrac-notification-engine-9   Running                                                                                                                                                                                                    0.0s 
 ✔ Container plextrac-notification-engine-10  Running                                                                                                                                                                                                    0.0s 
    docker-compose flags: ps --quiet notification-engine
    docker-compose configs:  -f /opt/plextrac/docker-compose.yml -f /opt/plextrac/docker-compose.override.yml
!!! trace
 at: cleanup(), /opt/plextrac/.local/bin/plextrac, line 33
  at: mod_rollout(), /opt/plextrac/.local/bin/plextrac, line 1738
   at: main(), /opt/plextrac/.local/bin/plextrac, line 195
    at: main(), /opt/plextrac/.local/bin/plextrac, line 2064
!!! trace
 at: cleanup(), /opt/plextrac/.local/bin/plextrac, line 33
  at: mod_rollout(), /opt/plextrac/.local/bin/plextrac, line 1738
   at: main(), /opt/plextrac/.local/bin/plextrac, line 195
    at: main(), /opt/plextrac/.local/bin/plextrac, line 2064
```

Reply:
yeah; I'm guessing I need to engineer a level-set to fix it at the start of the rollout to remove duplicate containers to fix it before scaling correctly. I'll work on this

Reply:
I fixed this; doing the below scales the services correctly before attempting a rollout. It takes the scale information from the `docker-compose*` files and removes containers outside that number or spins up new containers without recreation. You'd have to _really really_ work at it to get _some_ new containers with the old ones but this will also handle that too.
```bash
debug "$service_list"
  for s in ${service_list[@]}
    do
      debug "Stabalizing Service: $s"
      debug "`compose_client up -d --no-recreate $s > /dev/null 2>&1`"
  done
```

Result:
```text
Stabilizing Service: datalake-maintainer
Stabilizing Service: notification-engine
Stabilizing Service: notification-sender
Stabilizing Service: plextracapi
```

Reply:
Nice fix!

Reply:
This worked well in my testing!